zookeeper-server/src/test/java/org/apache/zookeeper/server/quorum/LearnerSyncThrottlerTest.java

```diff
@@ -0,0 +1,169 @@
+    public static Collection syncTypes() {
+        return Arrays.asList(new Object[][]{
+                {LearnerSyncThrottler.SyncType.DIFF}, {LearnerSyncThrottler.SyncType.SNAP}});
+    }
+    @Test(expected = SyncThrottleException.class)
+    public void testTooManySyncsNonessential() throws Exception {
+        LearnerSyncThrottler throttler =
+                new LearnerSyncThrottler(5, syncType);
+        for (int i = 0; i < 6; i++) {
+            throttler.beginSync(false);
+        }
+    }
+
+    @Test(expected = SyncThrottleException.class)
+    public void testTooManySyncsEssential() throws Exception {
+        LearnerSyncThrottler throttler =
+                new LearnerSyncThrottler(5, syncType);
+        try {
+            for (int i = 0; i < 6; i++) {
+                throttler.beginSync(true);
+            }
+        }
+        catch (SyncThrottleException ex) {
+            Assert.fail("essential syncs should not be throttled");
+        }
+        throttler.endSync();
+        throttler.beginSync(false);
+    }
+
+    @Test
+    public void testNoThrottle() throws Exception {
+        LearnerSyncThrottler throttler =
+                new LearnerSyncThrottler(5, syncType);
+        try {
+            for (int i = 0; i < 6; i++) {
+                throttler.beginSync(true);
+            }
+        }
+        catch (SyncThrottleException ex) {
+            Assert.fail("essential syncs should not be throttled");
+        }
+        throttler.endSync();
+        for (int i = 0; i < 5; i++) {
+            throttler.endSync();
+            throttler.beginSync(false);
+        }
+        Assert.assertTrue("should get here without exception", true);
+    }
+
+    @Test
+    public void testTryWithResourceNoThrottle() throws Exception {
+        LearnerSyncThrottler throttler =
+                new LearnerSyncThrottler(1, syncType);
+        for (int i = 0; i < 3; i++) {
+            throttler.beginSync(false);
+            try {
+                Assert.assertEquals(1, throttler.getSyncInProgress());
+            } finally {
+                throttler.endSync();
+            }
+        }
+    }
+
+    @Test(expected = SyncThrottleException.class)
+    public void testTryWithResourceThrottle() throws Exception {
+        LearnerSyncThrottler throttler =
+                new LearnerSyncThrottler(1, syncType);
+        throttler.beginSync(true);
+        try {
+            throttler.beginSync(false);
+            try {
+                Assert.fail("shouldn't be able to have both syncs open");
+            } finally {
+                throttler.endSync();
+            }
+        } finally {
+            throttler.endSync();
+        }
+    }
+
+    @Test
+    public void testParallelNoThrottle() throws Exception {
+        final int numThreads = 50;
+
+        final LearnerSyncThrottler throttler =
+                new LearnerSyncThrottler(numThreads, syncType);
+        ExecutorService threadPool = Executors.newFixedThreadPool(numThreads);
```

Comment:
Please shutdown this pool in a finally block

Reply:
done